lib/encoding/yaml/parser.toit

```diff
@@ -0,0 +1,1689 @@
+
+      if NAN.contains value: return float.NAN
+
+      if value.starts-with "0x":
+        if as-int := int.parse --radix=16 value[2..] --on-error=(: null):
+          return as-int
+
+      if value.starts-with "0o":
+        if as-int := int.parse --radix=8 value[2..] --on-error=(: null):
+          return as-int
+
+    if value is List:
+      return value.map: | node/ValueNode_ | node.resolve
+
+    if value is Map:
+      map := {:}
+      value.do:| key/ValueNode_  value/ValueNode_| map[key.resolve] = value.resolve
+      return map
+
+    return value
+
+  is-valid-key -> string?:
+    if value == null:
+      return "NULL_KEYS_UNSUPPORTED"
+    if value is List:
+      return "LIST_KEYS_UNSUPPORTED"
+    if value is Map:
+      return "MAP_KEYS_UNSUPPORTED"
+    return null
+
+  // As a potential key is checked against 'is-valid-key' above this hash-code
+  // should always succeed.
+  hash-code:
+    return value.hash-code
+
+
+/** Holds infomration about the properties applye to a node. */
+class NodeProperty_:
+ tag/string?
+ anchor/string?
+ constructor .tag .anchor:
+
+/** A base class for PEG parsers. */
+abstract class PegParserBase_:
+  bytes_/ByteArray
+  offset_/int := 0
+
+  constructor .bytes_:
+
+  // As a peg grammar is top-down, the parser needs to be able to rollback
+  mark -> any:
+    return offset_
+
+  rollback mark -> none:
+    offset_ = mark
+
+  at-mark mark -> bool:
+    return offset_ == mark
+
+  /**
+  Runs the given $block and rolls back if the returned value is null.
+  If the block uses a non-local return, then this function does *not* roll back.
+  Example usage:
+    try-parse:
+      if production-one-result := production-one:
+        if production-two-result := production-two:
+          return true
+    If any of the two production returns false/null then a roll back is issues,
+    otherwise no roll back is issues as we sucessfully parsed the two concatenated productions.
+
+    Use this for concatenated productions separated by |.
+  */
+  try-parse [block] -> none:
+    rollback-mark := mark
+    block.call
+    rollback rollback-mark
+
+  /**
+  Evaluated the %block and rolls back. The $block can return any value,
+    and the method will return a boolified version of the return value.
+    null and false are considered false values, everything else will be true
+  */
+  lookahead [block] -> bool:
+    result/any := false
+    try-parse: result = block.call
+    return not (not result) // Boolify the result.
+
+  /**
+  Evaluates the given $block starting $rune-count amount of runes behind the current offset.
+  */
+  lookbehind rune-count [block] -> bool:
+    old-offset := offset_
+    try:
+      rune-count.repeat:
+        if offset_ == 0:
+          return false
+        offset_--
+        // Skip to the first byte of Unicode surrogates.
+        // Byte 2, 3, and 4 all start with bits 10.
+        // Since the input is a valid string we only need to check for offset_ > 0.
```

Comment:
Because if the first rune is a surrogate, you would index -1.

Reply:
But then it wouldn't be a valid string. No?
So the question is: do we check before-hand that the input is actually a valid string.
If we do, then the `offset_` check isn't necessary.
If we don't then we need to keep it (and I would add a comment explaining that this might be the case).

Reply:
Right.